### PR TITLE
ROC-2054 Enter pay by set date

### PR DESCRIPTION
**Pull request by rattur**
Components for CCJ enter pay by set date

### JIRA link ###

https://tools.hmcts.net/jira/browse/ROC-2054

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/test/app/forms/validation/validators/IsTodayOrInFutureConstraint.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        expect(constraint.validate(new LocalDate(inFuture.year(), 1, 1))).to.equal(true)
+      })
+
+      it('given today date', () => {
+        let now = moment()
+        expect(constraint.validate(new LocalDate(now.year(), now.month() + 1, now.date()))).to.equal(true)
+      })
+    })
+
+    describe('should return false when ', () => {
```
Comment (timja-kainos) [inline src/test/app/forms/validation/validators/IsTodayOrInFutureConstraint.ts]: what about when the date is in the past?

File: src/test/app/forms/validation/validators/IsTodayOrInFutureConstraint.ts [hunk 2]
```diff
@@ -31,6 +31,11 @@ describe('DateTodayOrInFutureConstraint', () => {
       it('given an invalid structure', () => {
         expect(constraint.validate({ a: 1, b: 1, c: 2000 })).to.equal(false)
       })
+
+      it('given an date in the past', () => {
+        let inPast = moment().subtract(10, 'years')
+        expect(constraint.validate(new LocalDate(inPast.year(), 8, 8))).to.equal(undefined)
```
Comment (yarpo) [inline src/test/app/forms/validation/validators/IsTodayOrInFutureConstraint.ts]: what does it validate?

File: src/test/app/forms/validation/validators/IsTodayOrInFutureConstraint.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        expect(constraint.validate(new LocalDate(now.year(), now.month() + 1, now.date()))).to.equal(true)
+      })
+    })
+
+    describe('should return false when ', () => {
+      it('given an invalid structure', () => {
+        expect(constraint.validate({ a: 1, b: 1, c: 2000 })).to.equal(false)
+      })
+
+      it('given an date in the past', () => {
```
Comment (timja-kainos) [inline src/test/app/forms/validation/validators/IsTodayOrInFutureConstraint.ts]: > given a date in the past

File: src/main/features/ccj/form/models/payBySetDate.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  static readonly DATE_TODAY_OR_IN_FUTURE: string = 'Enter a date that is today or in the future'
+  static readonly DATE_INVALID_YEAR: string = 'Enter a 4 digit year'
+
+}
+
+export class PayBySetDate implements Serializable <PayBySetDate> {
+
+  @ValidateNested()
+  @IsDefined({ message: ValidationErrors.DATE_REQUIRED })
+  @isValidYearFormat(4, { message: ValidationErrors.DATE_INVALID_YEAR })
```
Comment (yarpo) [inline src/main/features/ccj/form/models/payBySetDate.ts]: not covered in unit tests

Comment (yarpo) [reply src/main/features/ccj/form/models/payBySetDate.ts]: Please see src/main/app/forms/models/address.ts:21 ValidationConstants
for constants used in validation annotations.

File: src/main/features/ccj/routes/pay-by-set-date.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      const form: Form<PayBySetDate> = req.body
+      const user: User = res.locals.user
+      if (form.hasErrors()) {
+        renderView(form, res)
+      } else {
+        const { externalId } = req.params
+        user.ccjDraft.payBySetDate = form.model
+        await DraftCCJService.save(res, next)
+        res.redirect(Paths.repaymentPlanPage.uri.replace(':externalId', externalId))
+
```
Comment (yarpo) [inline src/main/features/ccj/routes/pay-by-set-date.ts]: useless line (very, very minor)

File: src/main/features/ccj/views/pay-by-set-date.njk [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        <fieldset class="form-group {% if error %} form-group-error {% endif %}">
+
+          <legend>
+              <span class="form-label-bold">
+                {{ t("When you want them to pay the amount") }}
+              </span>
+          </legend>
+
+        </fieldset>
+
```
Comment (yarpo) [inline src/main/features/ccj/views/pay-by-set-date.njk]: too many empty lines

File: src/main/app/forms/validation/validators/IsTodayOrInFuture.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import {
```
Comment (timja-kainos) [inline src/main/app/forms/validation/validators/IsTodayOrInFuture.ts]: please ensure that check-and-send page is updated for this story, I've added a comment in the code where it needs to be changed and resolve the conflicts on this branch

Comment (timja-kainos) [reply src/main/app/forms/validation/validators/IsTodayOrInFuture.ts]: has been done

File: src/test/app/forms/validation/validators/IsTodayOrInFutureConstraint.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  describe('validate', () => {
+
+    describe('should return true when ', () => {
+      it('given a date is undefined or null', () => {
+        expect(constraint.validate(undefined)).to.equal(true)
+        expect(constraint.validate(null)).to.equal(true)
+      })
+
+      it('given a valid date in the future', () => {
+        let inFuture = moment().add(10, 'years')
```
Comment (timja-kainos) [inline src/test/app/forms/validation/validators/IsTodayOrInFutureConstraint.ts]: all these lets should be const but lets get this in...

**Pull request merged**